### PR TITLE
Add authn/authz to OTLP endpoint, refactor dashboard endpoint creation, integration tests

### PR DESCRIPTION
**Pull request by JamesNK**
Lots of dashboard changes for the reasons of auth. Work in progress.

* Change dashboard endpoint configuration from code to configuration + code. Allows [external endpoint configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-8.0#configure-endpoints-in-appsettingsjson) when starting the dashboard. This gives people who want to launch the dashboard a lot more flexibility over endpoint configuration. For example, the TLS cert can be different per endpoint via configuration, `Kestrel:Endpoints:Otlp:Certification:...=xxx`.
* Enable auth on OTLP endpoints
  * Add composite auth
  * Add API key auth (enabled via config)
  * Add client cert auth (enabled via config)
  * Change connection validation to happen in auth instead of gRPC interceptor.
* Integration tests of all the scenarios (http vs https vs different certs vs mTLS vs api key auth)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2316)

### CODE REVIEW

File: src/Aspire.Dashboard/Authentication/ApiKeyAuthorizationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Authorization;
+
+namespace Aspire.Dashboard.Authentication;
+
+public class ApiKeyAuthorizationHandler : AuthorizationHandler<ApiKeyAuthorizationRequirement>
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Authentication/ApiKeyAuthorizationHandler.cs:8]: In ASP.NET Core auth parlance, the Authentication handler deals with taking data from the HttpContext and turning it into an [AuthenticationTicket](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticationticket?view=aspnetcore-8.0) (ClaimsPrincipal + AuthenticationProperties) and then AuhorizationHandlers should use that information to decide if the user is authorized.

So this should be implemented as an AuthenticationHandler a simple claims check in the `ApiKeyAuthorizationHandler`.

Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/ApiKeyAuthorizationHandler.cs]: Done.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 2]
```diff
@@ -252,7 +252,10 @@ static Func<IPEndPoint> CreateEndPointAccessor(ListenOptions options)
     private static void ConfigurationAuthentication(WebApplicationBuilder builder)
     {
         builder.Services
-            .AddAuthentication(CertificateAuthenticationDefaults.AuthenticationScheme)
+            .AddAuthentication(OtlpCompositeAuthenticationDefaults.AuthenticationScheme)
```
Comment (davidfowl) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:255]: Why not base the action scheme on a configuration entry?

Comment (davidfowl) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:255]: PS: You can dynamically choose schemes using a policy scheme.

File: src/Aspire.Dashboard/Authentication/OtlpCompositeAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+using System.Text.Encodings.Web;
+using Aspire.Dashboard.Authentication.OtlpApiKey;
+using Aspire.Dashboard.Authentication.OtlpConnection;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Certificate;
+using Microsoft.Extensions.Options;
+
+namespace Aspire.Dashboard.Authentication;
+
+public sealed class OtlpCompositeAuthenticationHandler : AuthenticationHandler<OtlpCompositeAuthenticationHandlerOptions>
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/OtlpCompositeAuthenticationHandler.cs:14]: Connection auth (i.e. the OTLP request was received on the OTLP port) will always be present.

If we don't want to have mTLS and API key at the same time, then choosing whether to run them can happen here.

[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/OtlpCompositeAuthenticationHandler.cs:14]: I've added a mode setting for OTLP auth: None, ApiKey, ClientCertificate

There is still a composite handler which uses the mode to call the other handlers.

**Comment by davidfowl** [comment]
Do we need to test this with anything?

**Comment by joperezr** [comment]
Friendly reminder that this PR is assigned to the Preview 4 milestone which is snapping in a couple of hours. Please ensure it is merged by 4pm PST in order to ensure the change will be part of Preview 4.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 4]
```diff
@@ -193,6 +251,54 @@ static Func<IPEndPoint> CreateEndPointAccessor(ListenOptions options)
         }
     }
 
+    private static void ConfigurationAuthentication(WebApplicationBuilder builder, DashboardConfiguration dashboardConfig)
```
Comment (mitchdenny) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:254]: nit: Method name seems strange here. Should it be `ConfigureAuthentication`.

**Comment by mitchdenny** [comment]
I might be blind, but I'm not seeing where we make use of these values when we launch the dashboard from the apphost? Is that going to be done separately?

**Comment by JamesNK** [comment]
They're mostly for cloud providers to opt-in to using. We might add API key auth in the local environment.

**Comment by JamesNK** [comment]
I think this PR is mergable as a starting point for dashboard auth. I think there will still be changes, especially to the OTLP API key auth based on feedback, but this is a good starting point.

I don't want to leave this changing around, building up merge conflicts.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 5]
```diff
@@ -232,32 +342,66 @@ public ValueTask DisposeAsync()
...
-                var isHttps = IsHttps(uri);
-                if (isHttps)
-                {
-                    options.UseHttps();
-                }
-                configureListenOptions?.Invoke(options, isHttps);
+                case OtlpAuthMode.None:
+                    break;
+                case OtlpAuthMode.ApiKey:
+                    otlpApiKey = configuration[DashboardOtlpApiKeyVariableName];
```
Comment (kvenkatrajan) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:376]: Similarly here do we to fork out logic to retreive apiKey for ACA scenario? How would we determine from the appModel whether the dashboard is hosted in ACA? Can be in a separate PR - just curious

**Review by kvenkatrajan — COMMENTED** [review]
Looks great! Thanks for providing tests.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 6]
```diff
@@ -3,24 +3,31 @@
...
 namespace Aspire.Dashboard;
 
 public class DashboardWebApplication : IAsyncDisposable
 {
-    private const string DashboardOtlpUrlVariableName = "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL";
-    private const string DashboardOtlpUrlDefaultValue = "http://localhost:18889";
-    private const string DashboardUrlVariableName = "ASPNETCORE_URLS";
-    private const string DashboardUrlDefaultValue = "http://localhost:18888";
+    internal const string DashboardOtlpAuthModeVariableName = "DOTNET_DASHBOARD_OTLP_AUTH_MODE";
+    internal const string DashboardOtlpApiKeyVariableName = "DOTNET_DASHBOARD_OTLP_API_KEY";
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:26]: Maybe. Will talk more with people to figure out exact requirements. But I don't think that blocks this PR. Merging something that isn't the final result is ok.

File: src/Aspire.Dashboard/DashboardWebApplication.cs (hunk 5 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:376]: Still to figure out.

**Comment by iphdav** [comment]
One question: Has the industry adopted **x-otlp-api-key** as the name of the HTTP Header for the API Key?  It is obviously easy to use any header when emitting the telemetry.  Should the Options class in the Dashboard app allow a different key http header name to be specified as well as the key's value?
Great work though, this will be great.

**Comment by JamesNK** [comment]
I made it up. If there is a standard then let me know.

**Comment by iphdav** [comment]
I have just adopted your new standard James :-)

**Comment by JamesNK** [comment]
I'm still looking for someone to review this. Please review this so I don't have to keep rebasing changes to avoid merge conflicts.

**Comment by davidfowl** [comment]
I'll review

**Review by drewnoakes — COMMENTED** [review]
Would be good to document the new options in the dashboard's `README.md` file.

Also, lots of new public types here. Can these be internal, so we aren't tied to back compat in future? I don't see why consumers would need to use these types.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 7]
```diff
@@ -221,6 +285,52 @@ static Func<EndpointInfo> CreateEndPointAccessor(ListenOptions options, bool isH
...
+                            new Claim(ClaimTypes.Name,
+                                context.ClientCertificate.Subject,
+                                ClaimValueTypes.String, context.Options.ClaimsIssuer)
+                        };
+
+                        context.Principal = new ClaimsPrincipal(new ClaimsIdentity(claims, context.Scheme.Name));
+                        context.Success();
+
+                        return Task.CompletedTask;
+                    }
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:324]: That's right. I was planning to get feedback whether more to lock down the client cert was required.

File: src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        if (string.IsNullOrEmpty(Options.OtlpApiKey))
+        {
+            throw new InvalidOperationException("OTLP API key is not configured.");
+        }
+
+        var apiKey = Context.Request.Headers[ApiKeyHeaderName].ToString();
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs:25]: It won't throw.

But I've updated it to make the failure reason clearer.

File: src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        if (Options.OtlpApiKey != apiKey)
+        {
+            return Task.FromResult(AuthenticateResult.Fail("Incoming API key doesn't match required API key."));
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs:28]: I don't think the user will see these.

File: tests/Shared/TestCertificates/eku.client.pfx
[the remark this answers is not in the document]
Comment (JamesNK) [reply tests/Shared/TestCertificates/eku.client.pfx:1]: I don't know what they are. The README points to the ASP.NET Core repo which is where it was copied from. If they're available, then they're in aspnetcore.

**Comment by JamesNK** [comment]
> Would be good to document the new options in the dashboard's `README.md` file.

That's the plan, but I'll wait for concensus about the changes before updating docs.

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 10]
```diff
@@ -221,6 +285,52 @@ static Func<EndpointInfo> CreateEndPointAccessor(ListenOptions options, bool isH
         }
     }
 
+    private static void ConfigureAuthentication(WebApplicationBuilder builder, DashboardConfiguration dashboardConfig)
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:291]: The configuration about what auth is required is provided when the dashboard starts up. The `OtlpCompositeAuthenticationHandler` gets called and decides what auth to run based on config.

File: Directory.Packages.props [hunk 11]
```diff
@@ -25,6 +25,7 @@
     <PackageVersion Include="Azure.ResourceManager.Resources" Version="1.8.0-alpha.20240222.2" />
     <PackageVersion Include="Azure.Provisioning" Version="1.0.0-alpha.20240305.3" />
     <!-- ASP.NET Core dependencies -->
+    <PackageVersion Include="Microsoft.AspNetCore.Authentication.Certificate" Version="8.0.2" />
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply Directory.Packages.props]: Repo isn't using 8.0.3 yet. But I've made this a variable so it will get updated with everything else.

File: src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs]: Aren't these returned to the client in cases of failure? If a user in another culture is trying to configure certs for their environment, I expect they'd want to see these messages in their language.

**Comment by JamesNK** [comment]
> I'll review

@davidfowl I'm interested in your feedback about how I changed the configuration of endpoints. The new system allows for basic configuration using `ASPNETCORE_URLS` + `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL` or using more advanced Kestrel config, e.g. `Kestrel::Endpoints::OtlpEndpoint::xxxx`.

File: src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/OtlpApiKey/OtlpApiKeyAuthenticationHandler.cs]: It's pretty unlikely a gRPC client will send `accept-language` header with an API call. I searched opentelemetry-dotnet repo to double check and it doesn't send it.

I don't think we should worry about it.

**Review by drewnoakes — APPROVED** [review]

File: src/Aspire.Dashboard/DashboardWebApplication.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:329]: The ability to lock down the cert should be supported (file a follow up).

**Comment by drewnoakes** [comment]
@JamesNK @davidfowl can we get this merged in please? I have more auth work to submit that'll conflict a little with this.

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 12]
```diff
@@ -212,6 +202,82 @@ public DashboardWebApplication(Action<WebApplicationBuilder>? configureBuilder =
...
+            if (protocols != null)
+            {
+                values[$"Kestrel:Endpoints:{endpointName}:Protocols"] = protocols.ToString();
+            }
+
+            if (requiredClientCertificate)
+            {
+                values[$"Kestrel:Endpoints:{endpointName}:ClientCertificateMode"] = ClientCertificateMode.RequireCertificate.ToString();
+            }
+        }
```
Comment (davidfowl) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:244]: My only question (non blocking) was why do this instead of using the API to create named endpoints?

**Comment by drewnoakes** [comment]
Pushed a fix to resolve a simple conflict in `eng/Versions.props`.

File: src/Aspire.Dashboard/DashboardWebApplication.cs (hunk 12 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:244]: > * Change dashboard endpoint configuration from code to configuration + code. Allows [external endpoint configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-8.0#configure-endpoints-in-appsettingsjson) when starting the dashboard. This gives people who want to launch the dashboard a lot more flexibility over endpoint configuration. For example, the TLS cert can be different per endpoint via configuration, `Kestrel:Endpoints:Otlp:Certification:...=xxx`.

Without this, we'd need to recreate a lot of configuration infrastructure ourselves, i.e. provide a way to configure the TLS cert for each endpoint from a cert store, or from a pdf, or from a cert+key, etc.

File: eng/Versions.props [hunk 13]
```diff
@@ -33,6 +33,7 @@
     <MicrosoftExtensionsLoggingAbstractionsPackageVersion>8.0.1</MicrosoftExtensionsLoggingAbstractionsPackageVersion>
     <MicrosoftExtensionsOptionsPackageVersion>8.0.2</MicrosoftExtensionsOptionsPackageVersion>
     <MicrosoftExtensionsPrimitivesPackageVersion>8.0.0</MicrosoftExtensionsPrimitivesPackageVersion>
+    <MicrosoftAspNetCoreAuthenticationCertificatePackageVersion>8.0.2</MicrosoftAspNetCoreAuthenticationCertificatePackageVersion>
```
Comment (JamesNK) [inline eng/Versions.props:36]:
```suggestion
    <MicrosoftAspNetCoreAuthenticationCertificatePackageVersion>8.0.3</MicrosoftAspNetCoreAuthenticationCertificatePackageVersion>
```

File: src/Aspire.Dashboard/DashboardWebApplication.cs (hunk 12 above)
Comment (davidfowl) [reply src/Aspire.Dashboard/DashboardWebApplication.cs:244]: I see
